server/mailet/mailets/src/main/java/org/apache/james/transport/util/SpecialAddressesUtils.java

```diff
@@ -0,0 +1,272 @@
+        return builder.build();
+    }
+
+    private Set<MailAddress> getReplyTosFromMail(Mail mail) {
+        try {
+            InternetAddress[] replyToArray = (InternetAddress[]) mail.getMessage().getReplyTo();
+            if (replyToArray == null || replyToArray.length == 0) {
+                return getSender(mail);
+            }
+            return getReplyTos(replyToArray);
+        } catch (MessagingException ae) {
+            mailet.log("Unable to parse the \"REPLY_TO\" header in the original message; ignoring.");
+            return ImmutableSet.of();
+        }
+    }
+
+    private Set<MailAddress> getSender(Mail mail) {
+        MailAddress sender = mail.getSender();
+        if (sender != null) {
+            return ImmutableSet.of(sender);
+        }
+        return ImmutableSet.of();
+    }
+
+    private Set<MailAddress> getReplyTos(InternetAddress[] replyToArray) {
+        ImmutableSet.Builder<MailAddress> builder = ImmutableSet.builder();
+        for (InternetAddress replyTo : replyToArray) {
+            try {
+                builder.add(new MailAddress(replyTo));
+            } catch (ParseException pe) {
+                mailet.log("Unable to parse a \"REPLY_TO\" header address in the original message: " + replyTo + "; ignoring.");
+            }
+        }
+        return builder.build();
+    }
+
+    /**
+     * Returns a new Collection built over <i>list</i> replacing special
+     * addresses with real <code>InternetAddress</code>-es.<br>
+     * Manages <code>SpecialAddress.SENDER</code>,
+     * <code>SpecialAddress.REVERSE_PATH</code>,
+     * <code>SpecialAddress.FROM</code>, <code>SpecialAddress.REPLY_TO</code>,
+     * <code>SpecialAddress.RECIPIENTS</code>, <code>SpecialAddress.TO</code>,
+     * <code>SpecialAddress.NULL</code> and
+     * <code>SpecialAddress.UNALTERED</code>.<br>
+     * <code>SpecialAddress.RECIPIENTS</code> is made equivalent to
+     * <code>SpecialAddress.TO</code>.<br>
+     * <code>SpecialAddress.FROM</code> uses the From header if available,
+     * otherwise the Sender header if available, otherwise the return-path.<br>
+     * <code>SpecialAddress.REPLY_TO</code> uses the ReplyTo header if
+     * available, otherwise the From header if available, otherwise the Sender
+     * header if available, otherwise the return-path.<br>
+     * <code>SpecialAddress.UNALTERED</code> is ignored.<br>
+     * Any other address is not replaced.<br>
+     */
+    public List<MailAddress> replaceInternetAddresses(Mail mailWithReplacementAddresses, List<InternetAddress> internetAddresses) throws MessagingException {
+        ImmutableList.Builder<MailAddress> builder = ImmutableList.builder();
+        for (InternetAddress internetAddress : internetAddresses) {
+            MailAddress mailAddress = new MailAddress(internetAddress);
+            if (!SpecialAddress.isSpecialAddress(mailAddress)) {
+                builder.add(new MailAddress(internetAddress));
+                continue;
```

Comment:
I'm really not fan of continue instruction, which clearly breaks instruction flow and makes the code hard to reason about.... 

I prefer nested if + method extractions...

Here you can have a method called for each address that returns an Optional of MailAddress. Then you can add it to the builder if not empty.